pypeit/barframe.py

```diff
@@ -0,0 +1,468 @@
+                                                         self.tslits_dict['slit_righ'],
+                                                         self.tslits_dict['nspec'],
+                                                         self.tslits_dict['nspat'])
+        else:
+            self.slitmask_science = None
+            self.shape_science = None
+            self.shape_bar = None
+            self.nslits = 0
+            self.slit_left = None
+            self.slit_righ = None
+            self.slitcen = None
+            self.slitmask = None
+            self.gpm = None
+            self.nonlinear_counts = None
+
+    def build_traces(self, show_peaks=False, show_trace=False, debug=False):
+        """
+        Main routine to generate the bar profile traces in all slits
+
+        Args:
+             show_peaks (bool, optional):
+               Generate QA showing peaks identified by object finding
+             show_trace (bool, optional):
+               Generate QA  showing traces identified. Requires an open ginga RC modules window
+             debug (bool, optional):
+
+        Returns:
+            dict:  self.bar_dict
+        """
+        bar_prof = dict({})
+        nslits = self.tslits_dict['slit_left'].shape[1]
+        # Prepare the plotting canvas
+        if show_trace:
+            self.show('image', image=self.msbar.image, chname='bar_traces', slits=True)
+        # Go through the slits
+        for sl in range(nslits):
+            specobj_dict = {'setup': "unknown", 'slitid': sl,
+                            'det': self.det, 'objtype': "bar_profile", 'pypeline': "MultiSlit"}
+            msgs.info("Fitting bar traces in slit {0:d}".format(sl))
+            bar_traces, _ = extract.objfind(
+                self.msbar.image, self.slitmask == sl,
+                self.tslits_dict['slit_left'][:, sl],
+                self.tslits_dict['slit_righ'][:, sl],
+                ir_redux=False, ncoeff=self.par['trace_npoly'],
+                specobj_dict=specobj_dict, sig_thresh=self.par['sig_thresh'],
+                show_peaks=show_peaks, show_fits=False,
+                trim_edg=self.par['trim_edge'],
+                cont_fit=False, npoly_cont=0,
+                nperslit=len(self.par['locations']))
+            if len(bar_traces) != len(self.par['locations']):
+                # Bar tracing has failed for this slit
+                msgs.warn("Bar tracing has failed on slit {0:d}".format(sl))
+            if show_trace:
+                self.show('overplot', chname='bar_traces', bar_traces=bar_traces, slits=False)
+            bar_prof['{0:d}'.format(sl)] = bar_traces.copy()
+
+        bar_dict = self.generate_dict(bar_prof)
+
+        # Steps
+        self.steps.append(inspect.stack()[0][3])
+
+        # Return
+        return bar_dict
+
+    def generate_dict(self, bar_prof):
+        """
+        Generate a dictionary containing all of the information from the profile fitting
```

Comment:
We are going to want to use a `DataContainer`  for this instead of a `dict`.
Just a head's up for now.

Reply:
OK - if you need me to make changes, or refer to another PR/branch, just let me know!